### ISSUE TITLE
[bug] __pycache__ presence violates cached package integrity

### ISSUE THREAD

**Issue by ViliusSutkus89** [conan-io/conan#16707]
### Describe the bug

OS: Ubuntu-22.04
Conan version: 2.5.0
Conan profile:
```
[settings]
arch=x86_64
build_type=RelWithDebInfo
compiler=clang
compiler.cppstd=20
compiler.libcxx=libstdc++11
compiler.version=15
os=Linux
[conf]
tools.build:compiler_executables={'c': 'clang-15', 'cpp': 'clang++-15'}
```

Hello,

I came across this issue where GLib package has some python scripts in res folder, which are executed by GLib conan tests. This execution creates `__pycache__` files, which aren't supposed to be packaged, but cache integrity checker still evaluates them and errors out, because they are not mentioned in the manifest file. Cache integrity is kind of important to me, because it is executed as part of `conan upload`.

Regards,
Vilius

### How to reproduce it

```shell
$ conan create recipes/glib/all/conanfile.py --version 2.78.3 --build missing
$ conan cache check-integrity glib

Found 1 pkg/version recipes matching glib in local cache
glib/2.78.3: Integrity checked: ok
ERROR: glib/2.78.3:7f28aa2046b112e6880a8ce24f0b337eba9f5e55: Manifest mismatch
ERROR: Folder: /home/runner/.conan2/p/b/glib500aee2c65c90/p
ERROR:     'res/glib-2.0/codegen/__pycache__/utils.cpython-312.pyc' (manifest: None, file: 11b0ae06f1cdfefbc5ce50724f84b998)
ERROR:     'res/glib-2.0/codegen/__pycache__/__init__.cpython-312.pyc' (manifest: None, file: 6332e31c47ed9672e6eeadf2ad32d2fb)
ERROR:     'res/glib-2.0/codegen/__pycache__/codegen_md.cpython-312.pyc' (manifest: None, file: 4ea4ae2cf7c05fb48c28249addaa6261)
ERROR:     'res/glib-2.0/codegen/__pycache__/codegen.cpython-312.pyc' (manifest: None, file: 5c9a01394688c58cabc7b6718b6796d7)
ERROR:     'res/glib-2.0/codegen/__pycache__/codegen_rst.cpython-312.pyc' (manifest: None, file: 30e8c339625a71f34303159988cdb6ad)
ERROR:     'res/glib-2.0/codegen/__pycache__/dbustypes.cpython-312.pyc' (manifest: None, file: b636ecfbcb1155c8f33dc055da1f775d)
ERROR:     'res/glib-2.0/codegen/__pycache__/config.cpython-312.pyc' (manifest: None, file: fa337bc9277fd63cd045362af3eb9945)
ERROR:     'res/glib-2.0/codegen/__pycache__/parser.cpython-312.pyc' (manifest: None, file: 02e5daeb7be2ee940cfad7b9760cebcf)
ERROR:     'res/glib-2.0/codegen/__pycache__/codegen_main.cpython-312.pyc' (manifest: None, file: 2a4daec49470b399b7ba70aeec4dc6d4)
ERROR:     'res/glib-2.0/codegen/__pycache__/codegen_docbook.cpython-312.pyc' (manifest: None, file: 538d97fe4b8866d1f52f593362f00fdc)
ERROR: There are corrupted artifacts, check the error logs
```

Issue isolated and reproduced in this GitHub actions run - https://github.com/ViliusSutkus89/conan-center-index/actions/runs/10031877077/job/27722820637

**Comment by ViliusSutkus89** [comment]
As a workaround I do `find ~/.conan2 -type d -name __pycache__ -exec rm -r {} +` before running `conan upload`

**Comment by memsharded** [comment]
Hi @ViliusSutkus89 

Thanks very much for your report.
We are aware of this issue, we are working in a new ``install()`` method functionality that will allow to solve it: https://github.com/conan-io/conan/pull/16646, please track it to see progress (it will automatically close this ticket if merged).

As other potential workaround in the meantime, the PYTHONDONTWRITEBYTECODE could help, see https://github.com/conan-io/conan-center-index/pull/21764/files

**Comment by ViliusSutkus89** [comment]
Hello @memsharded . We have a workaround, so there's no rush on my end. I'll agree that PYTHONDONTWRITEBYTECODE workaround is way nicer than executing `rm -r`'s, so I'll be doing that

**Comment by memsharded** [comment]
Great, thanks for the feedback!

**Comment by ViliusSutkus89** [comment]
Turns out I don't know how to apply the workaround correctly. Initially I thought I've applied it correctly, but then it didn't work and I've entirely disabled the `gdbus-codegen` call, so `__pycache__`'s aren't generated.

Anyway, could you help me out? I've looked all over the internet, but couldn't figure out.

In [glib test conanfile.py](https://github.com/conan-io/conan-center-index/blob/9635424382e78d28793555e2d55f7e3eafcc3436/recipes/glib/all/test_package/conanfile.py#L28C16-L28C61): `self.run("gdbus-codegen -h", env="conanrun")`

How do I add `PYTHONDONTWRITEBYTECODE` env var to `conanrun` environment?

(To be honest, I need to know how to set env vars in tests in some other package, but this one would have the same exact solution, so it's easier to ask here)

**Comment by memsharded** [comment]
Hi @ViliusSutkus89 

There might be some different approaches, depending on the scope, if want to modify the recipe or not, etc.:

- You can define env-vars in your profile ``[runenv]`` section. It is also possible to define env-vars varlues per package with patterns such as ``mypkg/*:MYVAR=...``
- In the recipe you can add in the ``generate()`` method an ``Environment`` object, define your variables and assign it to ``scope="run"`` https://docs.conan.io/2/reference/tools/env/environment.html#environment, so it is added to the ``conanrun`` script.

Please let me know if this helps

**Comment by ViliusSutkus89** [comment]
I've tried to do this in [glib's test conanfile](https://github.com/conan-io/conan-center-index/blob/master/recipes/glib/all/test_package/conanfile.py), but I'm doing something wrong:
```python
def generate(self):
    env = Environment()
    env.define('PYTHONDONTWRITEBYTECODE', '1')
    envvars = env.vars(self, scope="run")

def test(self):
    self.run('env | grep PYTHON', env="conanrun")
```

I think I'm supposed to save `envvars` object somewhere, but if I use something like `envvars.save_script("conanrun")`, I'll just overwrite something. Am I supposed to use env different than conanrun when calling `self.run()`?

I've tried runenv in the profile too, but it didn't make it to test conanfile either:
```
[runenv]
glib/*:PYTHONDONTWRITEBYTECODE=1
```

Thanks for helping me, btw

**Comment by memsharded** [comment]
Something like:
```python
def generate(self):
    env = Environment()
    env.define('PYTHONDONTWRITEBYTECODE', '1')
    envvars = env.vars(self, scope="run")
    envvars.savescript("myownname")
```

This will:
- create the ``myownname`` script with your environment variables
- Add the ``myownname`` script to the ``conanrun`` script, so the second one calls the former

The problem is that this approach only works for the consumer of the code, so it should be added to every package that uses the package creating the ``__pycache__`` folder. 
What the ``meson`` recipe does is to add a wrapper that wraps meson injecting this env-var, check: https://github.com/conan-io/conan-center-index/blob/5f260b5ba6285f84298890a24485bc970d2e13dd/recipes/meson/all/conanfile.py#L45

This is done in the ``package()`` method of the recipe, so callers don't call meson directly but through the wrapper that injects the env-var.

Another approach would be to try to propagate the env-var from the ``package_info()`` of the recipe to the consumers.

**Comment by ViliusSutkus89** [comment]
`envvars.savescript("myownname")` worked for me. Somewhy I've thought that I had to add `myownname` to `conanrun` somehow.

I need this functionality in one specific place, so I'm fine with modifying consumer code. Thanks, now I know how to mess with env vars :)

**Comment by memsharded** [comment]
Good to know, thanks for the feedback!

I'll leave the ticket open, it will be closed by https://github.com/conan-io/conan/pull/16646, as that will be an adequate solution.

**Comment by memsharded** [comment]
``finalize()`` method PR has been merged https://github.com/conan-io/conan/pull/16646,

It will be released in next Conan 2.7, feedback welcome (please create new tickets). Thanks!
